Add latency statistics accumulation to GlueHiveMetastore

alexjo2144 opened the pull request:
These are used by the JMX connector to present stats similar to the ThriftMetastoreStats: https://github.com/prestosql/presto/blob/master/presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreStats.java

Issue: https://github.com/prestosql/presto/issues/1294

~~There were a few calls I didn't add stats to because they do async api calls so the stats didn't seem to make as much sense. Those were~~
- ~~getPartitionsByNames~~
- ~~addPartitions~~

aalbu commented:
Could you note which calls you didn't record stats for?

aalbu commented:
I noticed that there aren't any tests for the other metastores, but could you try to add one that proves stats are being collected here?  Basically make a call and then select from the corresponding `jmx.current...` table to show that the call count has been incremented?

alexjo2144 commented:
Added one unit test which just looks at the stat accumulation, I'll try to add another integration test to test the hook in with JMX.

alexjo2144 commented:
I was able to make a full integration test between Glue and the JMX connector, but had to make a few changes. 

- Added jmx as a test dependency to presto-hive
- Added one entry to the main pom's `dependencyManagement` list. I'm not fully sure I understand what that does, but was needed for the first point.
- Allowed for a null metastore in the TestingHiveConnectorFactory
  - Needed this in order for the `hive.metastore` property to take effect

alexjo2144 commented on presto-hive/pom.xml:
```diff
@@ -328,6 +334,7 @@
                         <configuration>
                             <excludes>
                                 <exclude>**/TestHiveGlueMetastore.java</exclude>
+                                <exclude>**/TestGlueJMXStatistics.java</exclude>
```
Looks like there's a `test-hive-glue` profile that got dropped when we moved from travis to github workflows?

electrum replied:
These were never actually used

electrum commented:
Let’s not use the JMX connector for testing, as that adds complexity and is not necessary. Instead, we can simply ask the MBean server and access the stats directly.

electrum commented:
I think this is missing the Guice binding to export stats. For now, we can drop the integration tests. Verify manually that the stats are being exported correctly when you run the full server.

alexjo2144 commented:
Sounds good, thanks David. I rolled back the last few commits with the integration test attempts. Here's a screenshot from the cli querying the stats via the jmx connector:

![image](https://user-images.githubusercontent.com/7109804/73777970-61b27180-4758-11ea-8a75-0b7ce2efa142.png)

findepi commented:
> There were a few calls I didn't add stats to because they do async api calls so the stats didn't seem to make as much sense. Those were
> 
> * getPartitionsByNames
> * addPartitions

I think we want to track these too. 

eg for one `getPartitionsByNames` call, we do multiple `batchGetPartitionAsync` requests.
We could even track both, but I think we can start from tracking `getPartitionsByNames`  invocations first.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueHiveMetastore.java:
```diff
@@ -245,17 +256,19 @@ private static AWSCredentialsProvider getCustomAWSCredentialsProvider(String pro
-            List<String> databaseNames = new ArrayList<>();
-            String nextToken = null;
+            return stats.getGetAllDatabases().wrap(() -> {
+                List<String> databases = new ArrayList<>();
```
unrelated change (used to be `databaseNames`). This can stay, but generally please avoid unrelated changes

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueHiveMetastore.java:
```diff
@@ -822,27 +849,33 @@ public void dropColumn(HiveIdentity identity, String databaseName, String tableN
+                        BatchCreatePartitionResult result = future.get();
+                        propagatePartitionErrorToPrestoException(databaseName, tableName, result.getErrors());
+                    }
+                    catch (InterruptedException e) {
```
Add `throw new PrestoException(HIVE_METASTORE_ERROR, e);` to catch clause (after re-interrupting current thread)

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueMetastoreApiStats.java:
```diff
@@ -0,0 +1,61 @@
+    private final TimeStat time = new TimeStat(MILLISECONDS);
+    private final CounterStat totalFailures = new CounterStat();
+
+    public <V, E extends Exception> V wrap(ThrowingCallable<V, E> callable)
```
"wrap" should return a wrapping callable (`Callable<V>`)

rename to `call`

(you can fix `JdbcApiStats` in separate PR)

findepi commented on presto-hive/src/test/java/io/prestosql/plugin/hive/metastore/glue/TestHiveGlueMetastore.java:
```diff
@@ -113,4 +115,27 @@ public void testGetPartitions() throws Exception
+        getMetastoreClient().getAllDatabases();
+        assertEquals(stats.getGetAllDatabases().getTime().getAllTime().getCount(), initialCallCount + 1.0);
+        assertTrue(stats.getGetAllDatabases().getTime().getAllTime().getAvg() > 0.0);
+        assertEquals(stats.getGetAllDatabases().getTotalFailures().getTotalCount(), initialFailureCount);
```
Add `@Test(singleThreaded = true)` to the test class

alexjo2144 commented:
Fixed and rebased. Thanks Piotr

findepi commented:
CI failed -- https://github.com/prestosql/presto/issues/2616

findepi commented:
Merged as 7c32534449ba2416876e059373e5852d817d893c, thanks!

(i amended the commit message)
